File: csharp/DataStructures/src/dsa/mathematics/RemoveEvenInteger.cs
namespace dsa.mathematics;

    public class RemoveEvenInteger
    {
        // static void Main()
        // {
        //     // Define um array de inteiros
        //     int[] arr = { 3, 2, 4, 7, 10, 6, 5 };
            
        //     // Chama o método PrintArray para imprimir o array
        //     PrintArray(arr);
            
        //     // Chama o método RemoveEven para remover os números pares do array
        //     int[] result = RemoveEven(arr);
            
        //     // Imprime o array resultante após a remoção dos números pares
        //     PrintArray(result);
        // }
        
        // Método para imprimir um array
        public static void PrintArray(int[] arr)
        {
            // Obtém o tamanho do array
            int n = arr.Length;
            
            // Loop para percorrer cada elemento do array
            for (int i = 0; i < n; i++)
            {
                // Imprime o elemento atual do array
                Console.Write(arr[i] + " ");
            }
            
            Console.WriteLine();
        }

        // Método para remover os números pares de um array
        public static int[] RemoveEven(int[] arr)
        {
            // Contador para os números ímpares
            int oddCount = 0;
            
            // Loop para percorrer cada elemento do array
            for (int i = 0; i < arr.Length; i++)
            {
                // Se o elemento atual é ímpar, incrementa o contador
                if (arr[i] % 2 != 0)
                    oddCount++;
            }
            
            // Cria um novo array para armazenar os números ímpares
            int[] result = new int[oddCount];
            
            // Índice para o novo array
            int idx = 0;
            
            // Loop para percorrer cada elemento do array original
            for (int i = 0; i < arr.Length; i++)
            {
                // Se o elemento atual é ímpar, adiciona ao novo array
                if (arr[i] % 2 != 0)
                {
                    result[idx] = arr[i];
                    idx++;
                }
            }
            
            // Retorna o novo array com apenas números ímpares
            return result;
        }
    }

File: go/dsa/mathematics/RemoveEvenInteger.go
package mathematics

import "fmt"

// PrintArray imprime os elementos de um array de inteiros
func PrintArray(array []int) {
	// Obtém o tamanho do array
	n := len(array)

	// Loop para percorrer cada elemento do array
	for i := 0; i < n; i++ {
		// Imprime o elemento atual do array
		fmt.Printf("%d ", array[i])
	}
	fmt.Println()
}

// RemoveEven remove os números pares de um array de inteiros e retorna um novo array com apenas números ímpares
func RemoveEven(array []int) []int {
	// Contador para os números ímpares
	oddCount := 0

	// Loop para percorrer cada elemento do array
	for i := 0; i < len(array); i++ {
		// Se o elemento atual é ímpar, incrementa o contador
		if array[i]%2 != 0 {
			oddCount++
		}
	}

	// Cria um novo slice para armazenar os números ímpares
	result := make([]int, oddCount)

	// Índice para o novo slice
	idx := 0

	// Loop para percorrer cada elemento do array original
	for i := 0; i < len(array); i++ {
		// Se o elemento atual é ímpar, adiciona ao novo slice
		if array[i]%2 != 0 {
			result[idx] = array[i]
			idx++
		}
	}

	// Retorna o novo slice com apenas números ímpares
	return result
}
